library/src/main/java/io/shortcut/firestoreextensions/FirestoreExtensions.kt

```
@@ -11,31 +14,116 @@ import com.google.firebase.firestore.MetadataChanges
     get(source).addOnSuccessListener { collection -> c.resume(collection) }
         .addOnFailureListener { e -> c.resumeWithException(e) }
 }
 
+/**
+ * Fetch the querySnapshot from cache ignoring exceptions.
+ */
+private suspend fun CollectionReference.fetchCacheOrNull() = suspendCoroutine<QuerySnapshot?> { c ->
+    this.get(Source.CACHE).addOnCompleteListener { task ->
+        c.resume(
+            if (task.isSuccessful) task.result else null
+        )
+    }
+}
+
+/**
+ * Returns a flow that stops after receiving a value from the server.
+ * Use this to read from cache and wait for a value from the server.
+ */
+private fun DocumentReference.flowOnce(): Flow<DocumentSnapshot?> = asFlow().transformWhile { snapshot ->
+    emit(snapshot)
+    snapshot.metadata.isFromCache
+}
+
+
+
 suspend fun Query.fetch(source: Source = Source.DEFAULT) = suspendCoroutine<QuerySnapshot> { c ->
     get(source).addOnSuccessListener { snapshot -> c.resume(snapshot) }
         .addOnFailureListener { e -> c.resumeWithException(e) }
 }
 
+private suspend fun Query.fetchOrNull(source: Source = Source.DEFAULT) = suspendCoroutine<QuerySnapshot?> { c ->
+    this.get(source).addOnCompleteListener { task ->
+        c.resume(
+            if (task.isSuccessful) task.result else null
+        )
+    }
+}
+
+/**
+ * Fetch the querySnapshot ignoring exceptions.
+ */
+private suspend fun Query.fetchCacheOrNull() = suspendCoroutine<QuerySnapshot?> { c ->
+    this.get(Source.CACHE).addOnCompleteListener { task ->
+        c.resume(
+            if (task.isSuccessful) task.result else null
+        )
+    }
+}
+
 suspend fun DocumentReference.fetch(source: Source = Source.DEFAULT) = suspendCoroutine<DocumentSnapshot> { c ->
     get(source).addOnSuccessListener { snapshot -> c.resume(snapshot) }
         .addOnFailureListener { e -> c.resumeWithException(e) }
 }
 
+/**
+ * Fetch the documentReference ignoring exceptions.
+ */
```

NIT: "Fetch the **DocumentSnapshot** ignoring exceptions"